changelog.d/14714.feature

```diff
@@ -0,0 +1 @@
+Add experimental support for [MSC3391](https://github.com/matrix-org/matrix-spec-proposals/pull/3391) (removing account data).
```

Comment:
> Code looks plausible over all. I'm a little skeptical about not deleting the data though (and filling it in with `{}`). #14244 doesn't really seem to address this approach at all and seems to document a completely different approach?
> 
> Would using `null` as the value for deleted items let us keep track of the stream ID while differentiating "empty" from "deleted"?

The advantage of using `{}` is that clients will then automatically sync the account data type with a value of `{}`, which as per MSC3391 implies to the client that the account data entry was deleted. It also comes with the benefit that older clients, which set account data content to `{}` to "delete" it, would actually start deleting items.

We could use `NULL`, but I don't see any benefit other than it being a more commonplace value for marking an item as deleted. Also note that the `content` column of `account_data` and `room_account_data` is currently marked as `NOT NULL`.

The reason for all of the faff with the `account_data_undelivered_deletes` table is the need to keep track of when we can remove the relevant row from `(room_)account_data` - after all devices of a user has seen the change.

Reply:
> The advantage of using `{}` is that clients will then automatically sync the account data type with a value of {`},` which as per [MSC3391](https://github.com/matrix-org/matrix-spec-proposals/pull/3391) implies to the client that the account data entry was deleted. It also comes with the benefit that older clients, which set account data content to `{}` to "delete" it, would actually start deleting items.

But we could still provide that to clients if we stored it as `NULL` by coercing it when passing back to clients.

> We could use `NULL`, but I don't see any benefit other than it being a more commonplace value for marking an item as deleted.

The main benefit is that we know whether the data was actually deleted or just cleared. I'm not sure if this difference actually matters or not though. Maybe we don't want to actually delete rows if the data is just cleared and not deleted?

> Also note that the content column of `account_data` and `room_account_data` is currently marked as `NOT NULL`.

I believe making a column non-null is not difficult: I don't think this would be a blocker.

> The reason for all of the faff with the `account_data_undelivered_deletes` table is the need to keep track of when we can remove the relevant row from `(room_)account_data` - after all devices of a user has seen the change.

Hmm, OK. I don't think I'm following all of the changes being made then. This PR Is only the first of a few which will make these changes, I guess I'm having trouble lining up what parts of #14244 are in this PR and what is a future PR. That has a bunch of things ticked off which I don't see in this PR?

Reply:
> The main benefit is that we know whether the data was actually deleted or just cleared. I'm not sure if this difference actually matters or not though. Maybe we don't want to actually delete rows if the data is just cleared and not deleted?

That's true, and could be beneficial in a somewhat niche use case. The plan https://github.com/matrix-org/synapse/issues/14244 talks about involves creating an entry in a `account_data_undelivered_deletes` table when a delete occurs. That has the side-effect of also letting us know which rows were deleted vs. set to `{}`.

After this PR, we don't need to differentiate between delete account data entries and those set to `{}` - as we should consider them the same thing. In either case, clients should see the entry as deleted. And we'll clear rows from the database via the information stored in the `account_data_undelivered_deletes` table.

Without the `account_data_undelivered_deletes` table (or some other method to track which devices still need to be informed about an account data entry being deleted) we can't delete rows from `(room_)account_data` anyway.

And in the case MSC3391 ends up not being accepted, leaving the rows as `{}` instead of NULL is probably more desirable.